### DISCUSSION TITLE
How to address state with optional fields when overwriting?

### DISCUSSION THREAD

**Discussion by thepuzzlemaster** [pmndrs/zustand#2610] [Q&A]
I have a state that looks like this:
```
type MyState = {
  optionalField?: string,
  requiredField: string
}
```
My actions look like this:
```
type MyActions = {
  setOptionalField: (value: string) => void
  setRequiredField: (value: string) => void
  setFields: (data: Partial<MyState>) => void
}
```
My default state looks like this:
```
const defaultInitState: MyState = {
  requiredField: 'someStringValue'
}
```
I'm using zustand in a nextjs application and my store looks like this:
```
export const createStore = (initState: MyState = defaultInitState) => {
  return createStore<MyStore>()((set) => ({
    ...initState,
    setOptionalField: (value) => set((state) => ({...state, optionalField: value}),
    setRequiredField: (value) => set((state) => ({...state, requiredField: value}),
    setFields: (value) => set(() => ({...defaultInitState, ...value}),
  }))
}
```

The issue is that I want to be able to clear out my state and reset it back to the defaultInitState with the potential for some overrides. Or another way to think about it, is that I want to be able to set 1 or more fields, and use the `defaultInitState` for any fields I don't include. But if I call `myState.setFields({})`, because of the shallow merge, it never overwrites any of the existing optional fields. If I change my `setFields()` action to use `replace`, then it overwrites all my actions.

Some possible options I've considered:
Separate the actions outside of the createStore function and then include them in my actual setter, but I both don't love the idea of that, and also haven't figured out if there's a syntax for that in typescript because `set` isn't available outside of the createStore context.
Another possibility is to make sure I explicitly include all fields in the `defaultInitState`, even ones which may have an undefined value. I don't love this idea because if I ever forget to include a field, its value won't be cleared out when calling `setFields` if it already has a value. I also don't love the idea of needing extra lines of code - unnecessarily.

Am I missing something obvious here? Or is there another potential approach I'm not thinking of? I've looked through a bunch of searches for this sort of thing, but haven't quite found a good answer.

**Comment by dai-shi** [comment]
That sounds a little bit tricky. I think you need `replace=true` and copy actions by yourself.

**Comment by thepuzzlemaster** [comment]
Thanks for the quick reply.

If there's no sort of built-in way to accomplish this, I did come up with the following solution which looks like it will work, based on your suggestion.

```
type GetMyActions = (
  set: (
    partial:
      | MyStore
      | Partial<MyStore>
      | ((state: MyStore) => MyStore | Partial<MyStore>),
    replace?: boolean | undefined
  ) => void
) => MyActions

const getMyActions: GetMyActions = (set) => ({
  setFields: (data) => set(() => ({ ...getMyActions(set), ...defaultInitState, ...data }), true),
  setOptionalField: (value) => set((state) => ({ ...state, optionalField: value })),
  setRequiredField: (value) => set((state) => ({ ...state, requiredField: value })),
})

export const createMyStore = (initState: UserState = defaultInitState) => {
  return createStore<MyStore>()((set) => ({
    ...initState,
    ...getMyActions(set),
  }))
}
```

I'm not sure if I like this more than just ensuring that my `defaultInitState` has every possible field present, but at least I have some options to consider here and talk through with my team.

Thanks.

**Reply by dai-shi** [reply]
Yeah, undefined-able field would be much easier:
```ts
type MyState = {
  optionalField: string | undefined,
  requiredField: string
}
```
